src/Simulation/Simulators/SparseSimulator/Native/quantum_state.hpp

```diff
@@ -0,0 +1,1506 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+#pragma once
+
+#include <string>
+#include <unordered_map>
+#include <random>
+#include <cmath>
+#include <functional>
+#include <algorithm>
+#include <list>
+#include <iostream>
+#include <atomic>
+#include <thread>
+#ifdef _OPENMP
+    #include <omp.h>
+#endif
+
+#include "basic_quantum_state.hpp"
+#include "quantum_hash_map.hpp"
+#include "types.h"
+#include "gates.h"
+
+using namespace std::literals::complex_literals;
+
+namespace Microsoft::Quantum::SPARSESIMULATOR
+{
+
+// power of square root of -1
+inline amplitude iExp(int power)
+{
+
+    int p = ((power % 4) + 8) % 4;
+    switch (p)
+    {
+    case 0:
+        return 1;
+    case 1:
+        return 1i;
+    case 2:
+        return -1;
+    case 3:
+        return -1i;
+    default:
+        return 0;
+    }
+    return 0;
+}
+
+template<size_t num_qubits>
+bool get_parity(std::bitset<num_qubits> bitstring){
+    return bitstring.count() % 2;
+}
+
+// Compares two bitsets as through they were bitstrings
+// Used to enforce an ordering on bitsets, though currently not referenced
+template<size_t N>
+inline bool operator<(const std::bitset<N>& lhs, const std::bitset<N>& rhs) {
+    std::bitset<N> mask = lhs ^ rhs;
+    std::bitset<N> const ull_mask = std::bitset<N>((unsigned long long) -1);
+    for (int i = static_cast<int>(N - 8*sizeof(unsigned long long)); i > 0; i-= static_cast<int>(8*sizeof(unsigned long long))){
+        if (((mask >> i) & ull_mask).to_ullong() > 0){
+            return ((lhs >> i) & ull_mask).to_ullong() < ((rhs >> i) & ull_mask).to_ullong();
+        }
+    }
+    return ((lhs) & ull_mask).to_ullong() < ((rhs) & ull_mask).to_ullong();
+}
+    
+// Transforms a vector of indices into a bitset where the indices indicate precisely
+// which bits are non-zero
+template<size_t num_qubits>
+std::bitset<num_qubits> get_mask(std::vector<logical_qubit_id> const& indices){
+    std::bitset<num_qubits> mask = std::bitset<num_qubits>();
+    for (logical_qubit_id index : indices) {
+        mask.set(index);
+    }
+    return mask;
+}
+
+template<size_t num_qubits>
+class QuantumState : public BasicQuantumState
```

Comment:
This file (and this class in particular) will definitely benefit from breaking-up into multiple files.
Or at least it would be good to separate the definition of the class from the definition of the member functions:
```c++
// Definition of the class:
template<size_t num_qubits>
class QuantumState : public BasicQuantumState
{
  public:
    QuantumState();  // The member functions are declared only, but not defined here.
    . . . 
  private:
    // Member variables.
};

// Definition of the member functions (if it is a template class then in the same .hpp file):
template<size_t num_qubits>
QuantumState::QuantumState() {
    . . .
}
```